src/webR/webr-worker.ts

```
@@ -108,29 +110,87 @@ function dispatch(msg: Message): void {
-              obj: {
-                type: capture.type(),
-                ptr: capture.ptr,
-                methods: RObject.getMethods(capture),
-              },
-              payloadType: 'ptr',
-            });
+
+            const prot = { n: 0 };
+
+            try {
+              const capture = captureR(data.code, data.env, data.options);
+              protectInc(capture, prot);
+
+              const result = capture.get('result');
+              const outputs = capture.get(2) as RList;
+
+              keep(result);
+
+              const n = outputs.length;
+              const output: any[] = [];
+
+              for (let i = 1; i < n + 1; ++i) {
+                const out = outputs.get(i);
+                const type = (out.pluck(1, 1) as RCharacter).toString();
+                const data = out.get(2);
+
+                if (type === 'stdout' || type === 'stderr') {
+                  const msg = (data as RString).toString();
+                  output.push({ type, data: msg });
+                } else {
+                  keep(data);
+                  const payload = {
+                    obj: {
+                      ptr: data.ptr,
+                      type: data.type(),
+                      methods: RObject.getMethods(data),
+                    },
+                    payloadType: 'ptr',
+                  } as WebRPayloadPtr;
+                  output.push({ type, data: payload });
+                }
+              }
+
+              const resultPayload = {
+                payloadType: 'ptr',
+                obj: {
+                  ptr: result.ptr,
+                  type: result.type(),
+                  methods: RObject.getMethods(result),
+                },
+              } as WebRPayloadPtr;
+
+              write({
+                payloadType: 'raw',
+                obj: {
+                  result: resultPayload,
+                  output: output,
+                },
+              });
+            } catch (_e) {
```

The entire `'request'` branch of the `switch` is now surrounded by a `try/catch` that returns thrown errors as a webR error payload, so I think you can just use `try/finally` here and let any errors bubble up and be caught by that instead.